File: javascript/util/shapes.js
function isPentagonal(num, pentagons) {
  if (pentagons && pentagons.length && pentagons.length >= num) {
    return pentagons[num];
  }

  return (Math.sqrt(24 * num + 1) + 1) % 6 === 0;
}

function getShapes(limit, getNthShapeFn) {
  const shapes = new Map();
  let n = 1;
  let term = 1;

  while (term < limit) {
    shapes.set(term, true);
    term = getNthShapeFn(n);

    n++;
  }

  return shapes;
}

const getNthTriangle = (n) => (n / 2) * (n + 1);
const getNthSquare = (n) => n * n;
const getNthPentagon = (n) => (n * (3 * n - 1)) / 2;
const getNthHexagon = (n) => n * (2 * n - 1);
const getNthHeptagon = (n) => (n * (5 * n - 3)) / 2;
const getNthOctagon = (n) => n * (3 * n - 2);

const getTriangles = (limit) => getShapes(limit, getNthTriangle);
const getSquares = (limit) => getShapes(limit, getNthSquare);
const getPentagons = (limit) => getShapes(limit, getNthPentagon);
const getHexagons = (limit) => getShapes(limit, getNthHexagon);
const getHeptagons = (limit) => getShapes(limit, getNthHeptagon);
const getOctagons = (limit) => getShapes(limit, getNthOctagon);

module.exports = {
  isPentagonal,
  getNthTriangle,
  getNthSquare,
  getNthPentagon,
  getNthHexagon,
  getNthHeptagon,
  getNthOctagon,
  getTriangles,
  getSquares,
  getPentagons,
  getHexagons,
  getHeptagons,
  getOctagons,
};

File: python/util/shapes.py
import math


def is_pentagonal(num, pentagons):
  if pentagons is not None and int(pentagons.keys()[-1]) >= num:
    return pentagons[num] == True

  return (math.sqrt(24 * num + 1) + 1) % 6 == 0


def get_shapes(limit, get_nth_shape_fn):
  shapes = {}
  n = 1
  term = 1

  while term < limit:
    shapes[term] = True
    term = get_nth_shape_fn(n)

    n += 1

  return shapes


get_nth_triangle = lambda n: (n / 2) * (n + 1)
get_nth_square = lambda n: n * n
get_nth_pentagon = lambda n: (n * (3 * n - 1)) / 2
get_nth_hexagon = lambda n: n * (2 * n - 1)
get_nth_heptagon = lambda n: (n * (5 * n - 3)) / 2
get_nth_octagon = lambda n: n * (3 * n - 2)

get_triangles = lambda limit: get_shapes(limit, get_nth_triangle)
get_squares = lambda limit: get_shapes(limit, get_nth_square)
get_pentagons = lambda limit: get_shapes(limit, get_nth_pentagon)
get_hexagons = lambda limit: get_shapes(limit, get_nth_hexagon)
get_heptagons = lambda limit: get_shapes(limit, get_nth_heptagon)
get_octagons = lambda limit: get_shapes(limit, get_nth_octagon)